cdc_iterator.go

```diff
@@ -0,0 +1,303 @@
+	case data := <-c.rowsEventsC:
+		rec, err := c.buildRecord(data)
+		if err != nil {
+			return rec, fmt.Errorf("failed to build record: %w", err)
+		}
+
+		return rec, nil
+	}
+}
+
+func (c *cdcIterator) Teardown(ctx context.Context) error {
+	close(c.canalDoneC)
+
+	c.canal.Close()
+	select {
+	case <-ctx.Done():
+		//nolint:wrapcheck // no need to wrap canceled error
+		return ctx.Err()
+	case err := <-c.canalRunErrC:
+		if err != nil {
+			return fmt.Errorf("failed to stop canal: %w", err)
+		}
+	}
+
+	return nil
+}
+
+func buildPayload(columns []schema.TableColumn, rows []any) sdk.StructuredData {
+	payload := sdk.StructuredData{}
+	for i, col := range columns {
+		payload[col.Name] = common.FormatValue(rows[i])
+	}
+	return payload
+}
+
+func (c *cdcIterator) buildRecord(e rowEvent) (sdk.Record, error) {
+	pos := common.CdcPosition{
+		Name: e.binlogName,
+		Pos:  e.Header.LogPos,
+	}
+
+	metadata := sdk.Metadata{}
+	metadata.SetCollection(e.Table.Name)
+	createdAt := time.Unix(int64(e.Header.Timestamp), 0).UTC()
+	metadata.SetCreatedAt(createdAt)
+
+	switch e.Action {
+	case canal.InsertAction:
+		position := pos.ToSDKPosition()
+		payload := buildPayload(e.Table.Columns, e.Rows[0])
+
+		table := common.TableName(e.Table.Name)
+		primaryKey := c.config.TableKeys[table]
+
+		key, err := buildRecordKey(primaryKey, payload)
+		if err != nil {
+			return sdk.Record{}, fmt.Errorf("failed to build record key: %w", err)
+		}
+
+		return sdk.Util.Source.NewRecordCreate(position, metadata, key, payload), nil
+	case canal.DeleteAction:
+		position := pos.ToSDKPosition()
+
+		payload := buildPayload(e.Table.Columns, e.Rows[0])
+
+		table := common.TableName(e.Table.Name)
+		primaryKey := c.config.TableKeys[table]
+
+		key, err := buildRecordKey(primaryKey, payload)
+		if err != nil {
+			return sdk.Record{}, fmt.Errorf("failed to build record key: %w", err)
+		}
+
+		return sdk.Util.Source.NewRecordDelete(position, metadata, key), nil
+	case canal.UpdateAction:
+		position := pos.ToSDKPosition()
+		before := buildPayload(e.Table.Columns, e.Rows[0])
+		after := buildPayload(e.Table.Columns, e.Rows[1])
+
+		table := common.TableName(e.Table.Name)
+		primaryKey := c.config.TableKeys[table]
+
+		key, err := buildRecordKey(primaryKey, before)
+		if err != nil {
+			return sdk.Record{}, fmt.Errorf("failed to build record key: %w", err)
+		}
+
+		return sdk.Util.Source.NewRecordUpdate(position, metadata, key, before, after), nil
+	}
+
+	return sdk.Record{}, fmt.Errorf("unknown row event action: %s", e.Action)
+}
+
+func buildRecordKey(
+	primaryKey common.PrimaryKeyName,
+	payload sdk.StructuredData,
+) (sdk.StructuredData, error) {
+	val, ok := payload[string(primaryKey)]
+	if !ok {
+		return nil, fmt.Errorf("key %s not found in payload", primaryKey)
+	}
+
+	return sdk.StructuredData{string(primaryKey): val}, nil
+}
+
+type rowEvent struct {
+	*canal.RowsEvent
+	binlogName string
+}
+
+type binlogName struct {
+	*sync.RWMutex
+	name string
+}
```

Comment:
Why do we need this? Do we expect the event handler methods to be called concurrently? That's not what I see when I look at the canal implementation. Even without looking at the implementation, I'd expect the events to be processed one after the other, otherwise ordering would be a pain and execution would be non-deterministic.